File: L1011-CapacityToShipPackagesWithinDDays.java
// https://leetcode.com/problems/capacity-to-ship-packages-within-d-days/
class Solution {
    public int[] getSum(int[] arr, int length) {
        int sum = 0, max = 0;
        for (int i = 0; i < length; i++) {
            sum += arr[i];
            if (max < arr[i]) {
                max = arr[i];
            }
        }
        return new int[] {max, sum};
    }
    
    public int getDays(int[] weights, int capacity, int length) {
        int days = 1, load = 0;
        for (int i = 0; i < length; i++) {
            if (load + weights[i] > capacity) {
                days = days + 1;
                load = weights[i];
            } else {
                load += weights[i];
            }
        }
        return days;
    }
    public int shipWithinDays(int[] weights, int days) {
        int length = weights.length;
        int t[] = getSum(weights, length);
        int low = t[0], high = t[1];
        while (low <= high) {
            int mid = (low + high) / 2;
            int day = getDays(weights, mid, length);
            if (day <= days) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return low;
    }
}

File: L1011-capacity_to_ship_packages_within_d_days.py
# https://leetcode.com/problems/capacity-to-ship-packages-within-d-days/
class Solution:
    def getMaxAndSum(self, arr: List[int], length: int) -> List[int]:
        s, m = 0 ,0
        for num in arr:
            s += num
            if (m < num):
                m = num
        return [m, s]
    
    def getDays(self, weights: List[int], length: int, capacity: int) -> int:
        days, load = 1, 0
        for weight in weights:
            if (load + weight > capacity):
                days = days + 1
                load = weight
            else:
                load += weight
        return days
    
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        length = len(weights)
        t = self.getMaxAndSum(weights, length)
        low, high = t[0], t[1]
        while (low <= high):
            mid = (low + high) // 2
            day = self.getDays(weights, length, mid)
            if (day <= days):
                high = mid - 1
            else:
                low = mid + 1
        return low
        
